CRM/MembershipExtras/BAO/MembershipPeriod.php

```diff
@@ -269,28 +265,94 @@ private static function doesOverlapWithOtherActivePeriods($periodParams) {
+    while ($membershipPeriods->N && $membershipPeriods->fetch()) {
+      $term ++;
+
+      if ($periodID && $membershipPeriods->id == $periodID) {
+        continue;
+      }
+
+      if (self::isStartDateOverlap($periodNewStartDate, $membershipPeriods)) {
+        $errors[] = "The new start date of this membership period overlaps with another activated membership period (term {$term}). Please review your changes.";
+      }
 
-    if (!empty($periodNewEndDate)) {
-      $overlappedMembershipPeriods->whereAdd('"' . $periodNewEndDate . '" >= start_date');
+      if (self::isEndDateOverlap($periodNewStartDate, $periodNewEndDate, $membershipPeriods)) {
+        $errors[] = "The new end date of this membership period overlaps with another activated membership period (term {$term}). Please review your changes.";
+      }
     }
 
-    $overlappedMembershipPeriods->whereAdd('id != ' . $periodID);
-    if ($overlappedMembershipPeriods->find()) {
-      return TRUE;
+    if (count($errors) > 0) {
+      $errors = implode(' ', $errors);
+      throw new CRM_Core_Exception('Overlapping errors were found when updating period: ' . $errors);
     }
+  }
 
-    return FALSE;
+  /**
+   * Checks if given start date overlaps with the given period's end date.
+   *
+   * @param string $periodStartDate
+   * @param \CRM_MembershipExtras_BAO_MembershipPeriod $evaluatedPeriod
+   *
+   * @return bool
+   * @throws \Exception
+   */
+  private static function isStartDateOverlap(
+    $periodStartDate,
+    CRM_MembershipExtras_BAO_MembershipPeriod $evaluatedPeriod
+  ) {
+    $startDate = new DateTime($periodStartDate);
+    $evaluatedPeriodStartDate = new DateTime($evaluatedPeriod->start_date);
+
+    if (empty($evaluatedPeriod->end_date)) {
+      return $startDate >= $evaluatedPeriodStartDate;
+    }
+
+    $evaluatedPeriodEndDate = new DateTime($evaluatedPeriod->end_date);
+
+    return $startDate <= $evaluatedPeriodEndDate && $startDate >= $evaluatedPeriodStartDate;
+  }
+
+  /**
+   * Checks if the given end date overlaps with the given period's start date.
+   *
+   * @param string $periodStartDate
+   * @param string $periodEndDate
+   * @param \CRM_MembershipExtras_BAO_MembershipPeriod $evaluatedPeriod
+   *
+   * @return bool
+   * @throws \Exception
+   */
+  private static function isEndDateOverlap(
+    $periodStartDate,
+    $periodEndDate,
+    CRM_MembershipExtras_BAO_MembershipPeriod $evaluatedPeriod
+  ) {
+    $startDate = new DateTime($periodStartDate);
+    $evaluatedPeriodStartDate = new DateTime($evaluatedPeriod->start_date);
+
+    if (empty($periodEndDate)) {
+      return $evaluatedPeriodStartDate >= $startDate;
+    }
+
+    $endDate = new DateTime($periodEndDate);
+    if (empty($evaluatedPeriod->end_date)) {
+      return $endDate >= $evaluatedPeriodStartDate;
+    }
+
+    $evaluatedPeriodEndDate = new DateTime($evaluatedPeriod->end_date);
+
+    return $endDate >= $evaluatedPeriodStartDate && $endDate < $evaluatedPeriodEndDate;
```

Comment:
shouldn't 

`$endDate < $evaluatedPeriodEndDate;`

be

`$endDate <= $evaluatedPeriodEndDate;`